src/sinks/amqp/encoder.rs

```diff
@@ -0,0 +1,26 @@
+use crate::{event::Event, sinks::util::encoding::Encoder};
+use bytes::BytesMut;
+use std::io;
+use tokio_util::codec::Encoder as _;
+
+#[derive(Clone, Debug)]
+pub(super) struct AMQPEncoder {
+    pub(super) encoder: crate::codecs::Encoder<()>,
+    pub(super) transformer: crate::codecs::Transformer,
+}
+
+impl Encoder<Event> for AMQPEncoder {
+    fn encode_input(&self, mut input: Event, writer: &mut dyn io::Write) -> io::Result<usize> {
+        let mut body = BytesMut::new();
+        self.transformer.transform(&mut input);
+        let mut encoder = self.encoder.clone();
+        encoder
+            .encode(input, &mut body)
+            .map_err(|_| io::Error::new(io::ErrorKind::Other, "unable to encode"))?;
+
+        let body = body.freeze();
+        writer.write_all(body.as_ref())?;
```

Comment:
Not sure the timing will align, but I have a PR open that provides a helper function to emit an EventsDropped internal event if `write_all` fails.
If you push before me, maybe we can add a TODO here.

https://github.com/vectordotdev/vector/pull/14225